refactor: update experience API rfc (#11)
logto-io/rfcs: draft/0004-experience-api.md

Reviewer: any specific reason for the trailing dots?

<img width="626" alt="image" src="https://github.com/user-attachments/assets/b36721e2-91c8-42fd-b984-54e6960acf78">

<img width="451" alt="image" src="https://github.com/user-attachments/assets/f910e9d0-69c4-44bb-ac64-7dab8138744c">

and no need for "the" in the title as it can be omitted for a shorter length and pretty print

Author: updated

  | interactionEvent | InteractionEvent | The interaction event for the current interaction. | Yes      |

- `POST /api/experience/submit`

  The `submit` API is used to submit the user's interaction to the Logto.

  - All precondition checks will be performed before submitting the interaction. The interaction will be rejected if any of the checks fail.
  - The interaction must be `Identified`, `Verified`, and `Profile fulfilled` to be submitted successfully.
  - Once the interaction is submitted, the new user profile and MFA settings will be updated in the user database.

- `PUT /api/experience/interaction-event`

  Update the interaction event for the current interaction.

  - The interaction event can be updated during the interaction to change the user's interaction purpose with Logto.
  - Only `SignIn` and `Register` interaction events are allowed to be interchanged during the interaction.
  - The interaction status will be recalculated based on the updated interaction event.
  - A non-existing verified email/phone verification record can be used to directly create a new user account by changing the interaction event from `SignIn` to `Register`.

>   - A existing verified email/phone verification record can be used to directly sign-in to the user account by changing the interaction event from `Register` to `SignIn`.

Reviewer 2: ```suggestion
  - An existing verified email/phone verification record can be used to directly sign-in to the user account by changing the interaction event from `Register` to `SignIn`.
```

#### Social/SSO

- `POST /api/experience/verification/{social|sso}/:connectorId/authorization-uri`

  Generate an authorization URI for the user to authenticate with a third-party social or enterprise identity provider.

  - A social/SSO verification record will be created in the interaction.
  - The unique `verificationId` will be generated for the verification record.
  - A `redirectUrl` will be generated for the user to authenticate with the social/SSO provider.

- `POST /api/experience/verification/{social|sso}/:connectorId/verify`

  Authenticate the user with the social/SSO provider and identify the user with the authenticated social/SSO identity.

  - The social/SSO verification record will be verified in the interaction.
  - The `verificationId` is required to retrieve the social/SSO verification record from the interaction.
  - For Google one tap sign-in, the `verificationId` can be omitted. A new verification record will be created.

  Request body:

  | Field          | Type                   | Description                                | Required |
  | -------------- | ---------------------- | ------------------------------------------ | -------- |

>   | data           | Record<string, string> | Callback data from the social/SSO provider | true     |

Reviewer 3: In other tables, we use Yes/No to indicate whether a field is required. Should align them

>   | verificationId | string                 | The verification record id                 | NO       |

Reviewer 4: ```suggestion
  | Field          | Type                   | Description                                | Required |
  | -------------- | ---------------------- | ------------------------------------------ | -------- |
  | data           | Record<string, string> | Callback data from the social/SSO provider | Yes     |
  | verificationId | string                 | The verification record id                 | No       |
```

>   | verificationId | string                 | The verification record id                 | NO       |

Reviewer 3: `No` instead of `NO`?

Author: Aligned to NO

  - Only `username` is supported as the identifier of the new password identity for now.
  - Can only be used in the `Register` interaction event to create and identify a new user account.

  Request body:

  | Field      | Type       | Description            | Required |
  | ---------- | ---------- | ---------------------- | -------- |
  | password   | string     | The user's password.   | Yes      |
  | identifier | Identifier | The user's identifier. | Yes      |

### 3. Identification API

`POST /api/experience/identification`

The identification API is used to identify the user with the given identifier during the sign-in experience. For the register interaction, the user will be created in the user database once a verified identifier is provided.

Request body:

| Field              | Type    | Description                                                                                                                                         | Required |
| ------------------ | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |

> | verificationId     | string  | The verification record id that can be used to identify the user.                                                                                   | NO       |

Reviewer 4: ```suggestion
| verificationId     | string  | The verification record id that can be used to identify the user.                                                                                   | No       |
```

> | verificationId     | string  | The verification record id that can be used to identify the user.                                                                                   | NO       |

Reviewer 3: `No` instead of `NO`?

Author: Aligned to NO

- In `Register` interaction event, the profile API can be used to provide the necessary profile data to create a new user account before the user is identified.
- In `SignIn` interaction event, the profile API can be used to fulfill the additional profile data and MFA settings that are required by the sign-in experience after the user is `identified` and `verified`.
- In `ForgotPassword` interaction event, the profile API can be used to set a new password for the user after the user is identified.

- `POST /api/experience/profile`

  The profile API to update the user's profile by a given type and value. Based on the profile type, verification record maybe required to verify the profile data.

  Request body:

  | Field          | Type   | Description                                      | Required |
  | -------------- | ------ | ------------------------------------------------ | -------- |
  | type           | union  | The profile data type to update.                 | true     |
  | value          | string | The profile data value to update.                | false    |

>   | verificationId | string | The verification record id for the profile data. | false    |

Reviewer 4: ```suggestion
  | type           | union  | The profile data type to update.                 | Yes     |
  | value          | string | The profile data value to update.                | No    |
  | verificationId | string | The verification record id for the profile data. | No    |
```

  Supported request body collections:

  | Type     | verificationId | value |
  | -------- | -------------- | ----- |
  | email    | true           | false |
  | phone    | true           | false |
  | password | false          | true  |
  | username | false          | true  |
  | social   | true           | false |

- `PUT /api/experience/profile/password`

  The profile API to reset the user's password in the `ForgotPassword` interaction event.
  For the first time password fulfillment, in the `Register` or `SignIn` interaction event, use the above `POST /api/experience/profile` instead.

  Request body:

  | Field    | Type   | Description       | Required |
  | -------- | ------ | ----------------- | -------- |
  | password | string | The new password. | Yes      |

- `POST /api/experience/profile/mfa`

  The mfa profile API to bind the MFA method to the user's profile.

  Request body:

  | Field          | Type    | Description                            | Required |
  | -------------- | ------- | -------------------------------------- | -------- |

>   | type           | MFAType | The MFA type to update in the profile. | No       |

Reviewer 3: not sure whether the type should be `MfaType`, please double check

>   | verificationId | string  | The verification record id for the MFA | Yes      |

Reviewer 5: The description should both have `period` marks.

    L-->>C: 200 OK
    C->>L: POST /api/experience/identification (new-password-identity verificationId)
    L-->>C: 200 OK
    C->>L: POST /api/experience/verification/web-authn/registration
    L-->>C: 200 OK
    C->>L: POST /api/experience/verification/web-authn/verify
    L-->>C: 200 OK
    C->>L: POST /api/experience/profile/mfa (type: WebAuthn)
    L-->>C: 200 OK
    C->>L: POST /api/experience/submit
```

### 17. Register with email and identity exists with the same email, sign-in directly

```mermaid
sequenceDiagram
    participant C as Client
    participant L as Logto

    C->>L: PUT /api/experience (register)
    L-->>C: 200 OK
    C->>L: POST /api/experience/verification/verification-code
    L-->>C: 200 OK
    C->>L: POST /api/experience/verification/verification-code/verify
    L-->>C: 200 OK
    C->>L: POST /api/experience/identification (email verificationId)
    L-->>C: 422 Account Already Exists
    C->>L: PUT /api/experience/interaction-event (sign-in)
    L-->>C: 200 OK
    C->>L: POST /api/experience/identification (email verificationId)
    L-->>C: 200 OK
    C->>L: POST /api/experience/submit
```

> ### 19. Forgot password with email

Reviewer 5: Should be 18?
